botocore/operation.py

```diff
@@ -77,6 +80,30 @@ def call(self, endpoint, **kwargs):
                           model=self.model,
                           params=kwargs)
         request_dict = self.build_parameters(**kwargs)
+
+        service_name = self.service.service_name
```

Comment:
The endpoint which is passed in contains no auth information. The endpoint heuristic resolver is used to load an endpoint config based on the passed in endpoint, which in turn may override auth information before we create the request signer. In this case the information in `endpoint_config` cannot be retrieved from `endpoint`, because it doesn't exist there anymore. Clients have the same logic in them.

The code is necessary because we need to know when to override the auth based on things like the hostname (e.g. switch to sigv4 for eu-central-1). The endpoint no longer has any associated auth information, so instead we do this after an endpoint has been created but before the request signer is created. The endpoint is still used to make the HTTP(S) request, but the request signer, which is created after checking for these overrides, is the one that signs the request before going out over the wire.